lib/dnscache.c

```diff
@@ -419,6 +419,31 @@ TLS_BLOCK_END;
 static DNSCacheOptions effective_dns_cache_options;
 G_LOCK_DEFINE_STATIC(unused_dns_caches);
 static GList *unused_dns_caches;
+volatile gint unused_dns_caches_ref_count;
+
```

Comment:
I think it would make sense to encapsulate this global state into a struct and add the refcounter for that struct.

Reply:
e.g. "DNSCaching" could be a name for this struct, as the functions use this as a prefix.